### PR TITLE
Keep older ledger snapshots

### PR DESCRIPTION
**Pull request by kderme**
https://github.com/input-output-hk/ouroboros-network/issues/2479

That's at a very early state, but some remarks so far:
- The filename for Genesis  is a special case. I'm using "genesis_..".
- If a slot already exists when we try to write a snapshot, we now just avoid writing it again. I think this is what this comment is refering to:
`-- TODO avoid taking multiple snapshots corresponding to the same tip.`

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 1]
```diff
@@ -317,13 +324,18 @@ trimSnapshots tracer hasFS DiskPolicy{..} = do
 -------------------------------------------------------------------------------}
 
 -- | On disk snapshots are numbered monotonically
-newtype DiskSnapshot = DiskSnapshot Int
+data DiskSnapshot = DiskSnapshot {
+    dsSlotNo :: !(WithOrigin SlotNo)
+  , dsSuffix :: !(Maybe String)
+  }
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:330]: Let's document it:
```suggestion
data DiskSnapshot = DiskSnapshot {
      -- | We name snapshots after the slot number the ledger state corresponds
      -- to. This gives an indication of how recent the snapshot is.
      --
      -- Note that the snapshot names are only indicative, we don't rely on them
      -- having the correct slot number. We only use the names to determine the
      -- order in which we try them. If a user chooses to rename their
      -- snapshots, we'll just try them in a different order. That order might
      -- be suboptimal, i.e., trying older snapshots requiring more replay
      -- first.
      dsSlotNo :: SlotNo
      -- | Snapshots can optionally have a suffix, separated by the slot number
      -- with an underscore, e.g., @4492799_last_Byron@. This suffix acts as
      -- metadata for the operator of the node. Snapshots with a suffix will
      -- /not be trimmed/.
    , dsSuffix :: Maybe String
    }
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 2]
```diff
@@ -317,13 +324,18 @@ trimSnapshots tracer hasFS DiskPolicy{..} = do
 -------------------------------------------------------------------------------}
 
 -- | On disk snapshots are numbered monotonically
-newtype DiskSnapshot = DiskSnapshot Int
+data DiskSnapshot = DiskSnapshot {
+    dsSlotNo :: !(WithOrigin SlotNo)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:328]: There's no point in making a snapshot of the ledger state at genesis, we are given this state at startup.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 3]
```diff
@@ -353,24 +365,43 @@ writeSnapshot hasFS encLedger encRef ss cs = do
...
+    pref = case dsSlotNo of
+      Origin       -> "genesis"
+      NotOrigin sl -> show $ unSlotNo sl
+    suff = case dsSuffix of
+      Nothing  -> ""
+      Just str -> "_" ++ str
 
 snapshotFromPath :: String -> Maybe DiskSnapshot
-snapshotFromPath = fmap DiskSnapshot . readMaybe
+snapshotFromPath fileName = fmap (flip DiskSnapshot fixedSuff) (readSlot pref)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:392]: Using `do` and record fields might be more readable here.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 4]
```diff
@@ -276,24 +276,31 @@ initStartingWith tracer conf streamAPI initDb = do
...
              -> HasFS m h
              -> (l -> Encoding)
              -> (r -> Encoding)
-             -> LedgerDB l r -> m (DiskSnapshot, WithOrigin r)
-takeSnapshot tracer hasFS encLedger encRef db = do
-    ss <- nextAvailable <$> listSnapshots hasFS
-    writeSnapshot hasFS encLedger encRef ss oldest
-    traceWith tracer $ TookSnapshot ss (csTip oldest)
-    return (ss, csTip oldest)
+             -> (r -> SlotNo)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:288]: Hmm, this is a sign of a leaky abstraction. The `LedgerDB` is parametric in `r`, yet here we say that it must have a `SlotNo`.

I think it would be slightly better if we turn this into `(r -> DiskSnapshot)`. This will also make it easier to later automatically (vs. manually by the user) give a suffix to snapshots. This also simplifies `mkIfAvailable` (I think it can go): if the `DiskSnapshot` for `tip` is in the list of snapshots, don't create one. When a named snapshot with the same slot number already exists, so be it, we'll later trim the one we created away anyway.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 5]
```diff
@@ -276,24 +276,31 @@ initStartingWith tracer conf streamAPI initDb = do
...
+      Nothing -> return (Nothing, csTip oldest)
+      Just ss -> do
+        writeSnapshot hasFS encLedger encRef ss oldest
+        traceWith tracer $ TookSnapshot ss (csTip oldest)
+        return (Just ss, csTip oldest)
   where
     oldest :: ChainSummary l r
     oldest = ledgerDbAnchor db
 
+    tip = csTip oldest
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:302]: BTW, I see 3 other uses of `csTip oldest` in the function above.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 6]
```diff
@@ -353,24 +365,43 @@ writeSnapshot hasFS encLedger encRef ss cs = do
     encode :: ChainSummary l r -> Encoding
     encode = encodeChainSummary encLedger encRef
 
--- | Delete snapshot from disk
-deleteSnapshot :: HasCallStack => HasFS m h -> DiskSnapshot -> m ()
-deleteSnapshot HasFS{..} = removeFile . snapshotToPath
+-- | Delete snapshot from disk, unless it has a suffix
+deleteSnapshot :: (HasCallStack, Monad m) => HasFS m h -> DiskSnapshot -> m ()
+deleteSnapshot HasFS{..} ds@DiskSnapshot {..} =
+    when (isNothing dsSuffix) $ removeFile $ snapshotToPath ds
```
[the remark this answers is not in the document]
Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:371]: We also call it when we fail to use a snapshot at `tryNewestFirst`. Do we still want to delete it in this case case, even though it has a suffix?

[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:371]: I see. I don't think we should delete it in that case. I suppose we can do a similar check in `tryNewestFirst`.

It *is* important to do the check in `trimSnapshots`, because I think it can change which snapshots get deleted.

[the remark this answers is not in the document]
Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:371]: > It is important to do the check in trimSnapshots, because I think it can change which snapshots get deleted.

I don't understand this part

[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs]: For example, say you want to keep 2 *non-named* snapshots on disk. You already have two named snapshots that are newer than all non-named snapshot. The LedgerDB just created its third non-named snapshot 140 and now trims it snapshots:
```
1000_shelley
200_favourite_snapshot
140
120
100
```
In `trimSnapshots`, you'll find there are 5 snapshots on disk, now you  drop the first two and try to delete the remaining three: `140`, `120`, and `100`. You end up without non-named snapshots instead of 2.

If you filter first in `trimSnapshots` on "non-named", you'll drop `140` and `120` from the list and delete `100`. Now you end up with 2 non-named snapshots.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 7]
```diff
@@ -353,24 +365,43 @@ writeSnapshot hasFS encLedger encRef ss cs = do
...
 
 -- | List on-disk snapshots, most recent first
 listSnapshots :: Monad m => HasFS m h -> m [DiskSnapshot]
 listSnapshots HasFS{..} =
     aux <$> listDirectory (mkFsPath [])
   where
     aux :: Set String -> [DiskSnapshot]
-    aux = List.sortBy (flip compare) . mapMaybe snapshotFromPath . Set.toList
+    aux = List.sortBy (flip compare) . mapMaybe snapshotFromPath  . Set.toList
```
[the remark this answers is not in the document]
Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs]: I noticed that this changed a bit which snapshots are deleted. ie if we have
100
100_a
200

previously it would compare based on the Ord instance of `DiskSnapshot`  (not just the slot) and it would delete the 100. now since they have the same slot, it's undeterministic (on my manual tests it deletes nothing).

[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs]: We can change the `Ord` instance to prefer a named snapshot over an unnamed one with the same slot number (and document it!).

For deleting, we should filter out the named ones first as I said elsewhere, so it shouldn't play a role there.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs]: So we want `onDiskNumSnapshots` to indicate the number of  non-named snapshots? That's not how this pr works

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 8]
```diff
@@ -274,56 +278,88 @@ initStartingWith tracer conf streamAPI initDb = do
 -- testing purposes, 'takeSnapshot' returns the block reference corresponding
 -- to the snapshot that we wrote.
 --
--- NOTE: This is a lower-level API that unconditionally takes a snapshot
--- (i.e., independent from whether this snapshot corresponds to a state that
--- is more than @k@ back).
+-- NOTE: This is a lower-level API that takes a snapshot independent from
+-- whether this snapshot corresponds to a state that is more than @k@ back. It
+-- doesn't take a snapshot if a snapshot at the same slot already exists on
+-- disk or the tip is the genesis.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:284]:
```suggestion
-- disk or if the tip is at genesis.
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 9]
```diff
@@ -274,56 +278,88 @@ initStartingWith tracer conf streamAPI initDb = do
...
-    writeSnapshot hasFS encLedger encRef ss oldest
-    traceWith tracer $ TookSnapshot ss (csTip oldest)
-    return (ss, csTip oldest)
+             -> (r -> DiskSnapshot)
+             -> LedgerDB l r
+             -> m (Maybe DiskSnapshot, WithOrigin r)
+takeSnapshot tracer hasFS encLedger encRef toSnap db = case tip of
+    Origin      -> return (Nothing, tip)
+    NotOrigin r -> do
+      let snapShot = toSnap r
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:303]: For consistency:
```suggestion
      let snapshot = toSnap r
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 10]
```diff
@@ -274,56 +278,88 @@ initStartingWith tracer conf streamAPI initDb = do
...
 trimSnapshots :: Monad m
               => Tracer m (TraceEvent r)
               -> HasFS m h
               -> DiskPolicy
-              -> m [DiskSnapshot]
+              -> m ()
 trimSnapshots tracer hasFS DiskPolicy{..} = do
     snapshots <- listSnapshots hasFS
+    -- We ignore named snaspshots
+    let snapshots' = filter (isNothing . dsSuffix) snapshots
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:328]:
```suggestion
    -- We ignore named snapshots
    snapshots <- filter (isNothing . dsSuffix) <$> listSnapshots hasFS
```
Now we can't accidentally refer to `snapshots` instead of `snapshots'`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 11]
```diff
@@ -274,56 +278,88 @@ initStartingWith tracer conf streamAPI initDb = do
...
---
--- The deleted snapshots are returned.
+-- snapshots are stored on disk. The oldest snapshots are deleted. Snapshots
+-- with a suffix are not deleted.
 trimSnapshots :: Monad m
               => Tracer m (TraceEvent r)
               -> HasFS m h
               -> DiskPolicy
-              -> m [DiskSnapshot]
+              -> m ()
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:324]: What's the reason for this change?

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 12]
```diff
@@ -357,20 +393,40 @@ writeSnapshot hasFS encLedger encRef ss cs = do
...
+    pref = show $ unSlotNo dsSlotNo
+    suff = case dsSuffix of
+      Nothing  -> ""
+      Just str -> "_" ++ str
 
 snapshotFromPath :: String -> Maybe DiskSnapshot
-snapshotFromPath = fmap DiskSnapshot . readMaybe
+snapshotFromPath fileName = do
+    slot <- readSlot pref
+    Just $ DiskSnapshot slot fixedSuff
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:419]: Embrace the monad :grin: 
```suggestion
    return $ DiskSnapshot slot fixedSuff
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs (hunk 11 above)
Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:324]: It wasn't used anywhere

**Comment by mrBliss** [comment]
I have rebased and changed this PR in #2733.
